docs/Port-Models.md

```diff
@@ -0,0 +1,318 @@
+
+Below is an example:
+
+```python
+@LmConfig.register_subclass("gpt2") # if implementing a Causal Language model
+@dataclass(frozen=True)  # for parsing the config file
+class MyConfig(HFCompatConfig):
+    # ...
+    @property
+    def model_type(cls) -> Type["Gpt2LMHeadModel"]:
+        return Gpt2LMHeadModel
+```
+
+### Write Model
+After you have defined your config class, you can start writing your model.
+You can follow the breakdown of the model in your Hugging Face implementation (or from other sources). This would make it easier to validate the implementation through unit tests.
+
+For example, in GPT2, we have the following breakdown:
+- `Gpt2Mlp`
+- `Gpt2Attention`
+- `Gpt2Block`: a block of Gpt2Mlp and Gpt2Attention
+- `Gpt2Transformer`: a stack of Gpt2Block
+- `Gpt2Embeddings`: token and position embeddings
+- `Gpt2LMHead`: a complete GPT2 model with embedding, transformer, and LM head.
+
+We follow the same breakdown in the implementation of Llama in Levanter.
+
+#### Note on the Implementation Format
+- Each class will have its key layers and components defined as attributes and be initialized with a static method `init()`.
+- If Each class will be extended from Equinox's `Module` class.
+
+### Serialization to/from State Dicts
+
+PyTorch and Hugging Face Transformers use "state dicts" as their preferred serialization format, either as pickles or as the new [safetensors](https://github.com/huggingface/safetensors) format.
+A state dict is a Python `dict` with string keys and tensor values. The keys of the dict are json-ish "key paths" like `model.blocks.0.mlp.c_proj` and the values are the corresponding parameters for that key path.
+You can read more about [PyTorch State Dicts here](https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html).
+
+Levanter has machinery for (de)serializing to and from state dicts. Simple cases are handled automatically, but often times custom logic is needed.
+
+#### Easy Case: Identical Module Structure
+If your module has exactly the same fields with the same names and same shapes as the Hugging Face state dict (e.g. Gpt2Mlp), you don't need to do anything.
+
+#### Medium Case: Different Names
+If for some reason you want to use different names from the HF implementation (e.g. because the names from HF aren't clear...), you can extend your class from  `StateDictSerializationMixin` and use `_state_dict_key_map` to rename keys. For instance, the `Gpt2Transformer` class has this method:
+
+```python
+class Gpt2Transformer(StateDictSerializationMixin, eqx.Module):
+    ...
+
+    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
+        return {"blocks": "h"}
+```
+
+This says that the field called `blocks`` in this class should be (de)serialized as `h``, because the Hugging Face GPT-2 implementation uses `h``, which is not very clear. You can also "flatten" the submodules of a field by using `None`` or even include `.s` in the name if needed.
```

Comment:
```suggestion
This says that the field called `blocks` in this class should be (de)serialized as `h`, because the Hugging Face GPT-2 implementation uses `h`, which is not very clear. You can also "flatten" the submodules of a field by using `None` or even include `.s` in the name if needed.
```